src/app/services/theme/theme.service.ts

```diff
@@ -0,0 +1,202 @@
+
+export interface RGB {
+  red: number;
+  green: number;
+  blue: number;
+}
+
+export interface HSL {
+  hue: number;
+  saturation: number;
+  lightness: number;
+}
+
+@Injectable({
+  providedIn: "root",
+})
+export class ThemeService {
+  public constructor(
+    @Inject(DOCUMENT) private document: Document,
+    @Inject(IS_SERVER_PLATFORM) private isServer: boolean
+  ) {}
+
+  /**
+   * Set the base colour for a theme
+   *
+   * @param theme Theme to modify
+   * @param color New base colour for theme
+   */
+  public setTheme(theme: BawTheme, color: string): void {
+    if (!this.validateHexRgb(color)) {
+      console.warn(`Invalid theme color given for ${theme} detected: ${color}`);
+      return;
+    }
+
+    const hsl = this.rgb2hsl(color);
+    this.updateTheme(
+      theme,
+      `
+        :root {
+          --baw-${theme}-hue: ${hsl.hue}deg;
+          --baw-${theme}-saturation: ${hsl.saturation}%;
+          --baw-${theme}-lightness: ${hsl.lightness}%;
+        }
+      `
+    );
+  }
+
+  /**
+   * Reset any modifications to the base colour of a theme
+   * ! This includes changes set by the environment
+   *
+   * @param theme Theme to reset
+   */
+  public resetTheme(theme: BawTheme): void {
+    const style = this.document.getElementById(this.getThemeId(theme));
+    if (style) {
+      this.document.head.removeChild(style);
+    }
+  }
+
+  /**
+   * Update a theme by either inserting, or modifying, a style in the header
+   *
+   * @param theme Theme to update
+   * @param styles CSS styles to apply
+   */
+  private updateTheme(theme: BawTheme, styles: string): void {
+    const id = this.getThemeId(theme);
+
+    // Get style element
+    let style = this.document.getElementById(id) as HTMLStyleElement;
+    if (style) {
+      style.innerHTML = styles;
+      return;
+    }
+
+    // Create style element if not exists
+    style = this.document.createElement("style");
+    style.id = id;
+    style.innerHTML = styles;
+    this.document.head.appendChild(style);
+  }
+
+  public extractHexRgb(hex: string): RGB {
+    const color: RGB = { red: 0, green: 0, blue: 0 };
+
+    /**
+     * Extract hex string and convert to int using base 16
+     */
+    function extractHex(...chars: [number, number]): number {
+      const value = hex.charAt(chars[0]) + hex.charAt(chars[1]);
+      return parseInt(value, 16);
+    }
+
+    if (hex.length === 7) {
+      color.red = extractHex(1, 2);
+      color.green = extractHex(3, 4);
+      color.blue = extractHex(5, 6);
+    } else {
+      color.red = extractHex(1, 1);
+      color.green = extractHex(2, 2);
+      color.blue = extractHex(3, 3);
+    }
+
+    return color;
+  }
+
+  /**
+   * Convert rgb hex string into an object containing the hue, saturation and
+   * lightness of the colour
+   *
+   * @param color RGB hex string to convert
+   * @returns HSL colour equivalent to rgb
+   */
+  private rgb2hsl(color: string): HSL {
```

Comment:
Really don't want to see this function in our code. External impl available?

Reply:
I was under the impression you disliked colour libraries because of the bulk they add. I'll attempt to find a lightweight library

Reply:
as long as we have tree shaking, we should be able to import just the functions we need, without the full library

Reply:
last i checked the latest d3 libs are ultra portable and we can import singular functions

https://github.com/d3/d3-color